rfcs/0087-aarch64-tier1.md

```diff
@@ -0,0 +1,93 @@
+as an additional result of channel bumps waiting for aarch64 builds to finish,
+saving on build times for end users.
+
+## Prior art
+There were prior attempts at the same feat, but they failed due to technical
+limitations of Hydra:
+ - NixOS/nixpkgs@74c4e30 - disabled in 2017 because of memory issues
+ - NixOS/nixpkgs#52534, NixOS/nixpkgs@36a0c13 - re-enabled in 2018 to pre-build important outputs
+ - NixOS/nixpkgs@1bfe8f1 - disabled again due to hydra-evaluator issues
+
+# Detailed design
+[design]: #detailed-design
+
+If this RFC is accepted, `aarch64-linux` builds will be added to stable
+and unstable channels' `tested` aggregate jobs on Hydra, giving them ability
+to block channel advances. Hydra will start building aarch64 packages and run
+aarch64-based tests as part of stable and unstable channels, including them in
+the binary cache, increasing its coverage as a result.
+
+It is possible that the availability of aarch64 builders from Equinix Metal will at times be
+reduced, causing delays in aarch64 build capacity. We will extend the
+nixos-org-configurations implementation of hydra-provisioner to dynamically allocate
+aarch64 builders on AWS during these capacity shortfalls.
+
+# Examples and Interactions
+[examples-and-interactions]: #examples-and-interactions
+
+<!-- This section illustrates the detailed design. This section should clarify all
+confusion the reader has from the previous sections. It is especially important
+to counterbalance the desired terseness of the detailed design; if you feel
+your detailed design is rudely short, consider making this section longer
+instead. -->
+
+In [nixos/release-combined.nix](https://github.com/NixOS/nixpkgs/blob/master/nixos/release-combined.nix)
+`aarch64-linux` will be moved to `supportedSystems`, enabling NixOS tests
+to block channel advances in case of failures.
+
+Merging this RFC should happen simultaneously with the merging of documentation
+around configuring qemu-binfmt as a fallback method for building aarch64 packages on
+x86_64 machines. Additionally, a sub-project that's out-of-scope for this RFC may be
+established to catch build failures (of which sightings were reported) when using
+emulation.
+
+The list of NixOS AMIs on NixOS.org will also be extended to include aarch64 images.
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+Some build failures could unneccesarily delay channel advances, delaying critical updates.
+
+# Alternatives
+[alternatives]: #alternatives
+
+Create an aarch64-focused channel that would build same things current `unstable` does, but for aarch64 only. This has a significant drawback: it is possible for the x86_64 channel and the aarch64 channels to never pass on the same commit, making deployment to a heterogeneous cluster of x86_64 and aarch64 machines very challenging.
+
+# Unresolved questions
+[unresolved]: #unresolved-questions
+
+Do we have enough machines to handle aarch64 builds without delaying `x86_64-linux` builds?
```

Comment:
Lately aarch64 has kept up fine with x64-linux. However, if we promote aarch64 some extra capacity might be needed for jobs that test larger changes. Often only x86_64 is included in those types of jobs.

Reply:
Nixpkgs-unstable already ends up trying to build everything NixOS would for aarch64-linux. I'm not sure this would change anything *except* for the staging-next workflow.

There's not much world rebuilds with the stable channels, so I would estimate them as not being an issue. I could be wrong.

Reply:
For extra capacity needs, see Detailed Design, paragraph 2.